src/util/wacz.ts

```diff
@@ -0,0 +1,424 @@
+    this.pagesDir = config.pages;
+    this.logsDir = config.logDirectory;
+    this.tempCdxDir = config.tempCdxDir;
+    this.collDir = collDir;
+    this.indexesDir = config.indexesDir;
+
+    this.datapackage = {
+      resources: [],
+      // drop microseconds
+      created: new Date().toISOString().split(".", 1)[0] + "Z",
+      wacz_version: "1.1.1",
+      software: config.softwareString,
+    };
+
+    if (config.title) {
+      this.datapackage.title = config.title;
+    }
+    if (config.description) {
+      this.datapackage.description = config.description;
+    }
+
+    this.signingUrl = config.signingUrl || null;
+    this.signingToken = config.signingToken || null;
+  }
+
+  generate(): Readable {
+    const files = [
+      ...this.warcs,
+      ...addDirFiles(this.indexesDir),
+      ...addDirFiles(this.pagesDir),
+      ...addDirFiles(this.logsDir),
+    ];
+
+    const zip = makeZip(
+      this.iterDirForZip(files),
+    ) as ReadableStream<Uint8Array>;
+
+    const hasher = createHash("sha256");
+    const resources = this.datapackage.resources;
+
+    let size = 0;
+
+    async function* iterWACZ(wacz: WACZ): AsyncIterable<Uint8Array> {
+      let isInFile = false;
+
+      let currMarker: StartMarker | null = null;
+
+      for await (const chunk of zip) {
+        if (chunk instanceof StartMarker) {
+          isInFile = true;
+          currMarker = chunk;
+        } else if (chunk instanceof EndMarker) {
+          isInFile = false;
+          if (currMarker) {
+            // Frictionless data validation requires this to be lowercase
+            const name = basename(currMarker.filename).toLowerCase();
+            const path = currMarker.zipPath;
+            const bytes = currMarker.size;
+            const hash = "sha256:" + currMarker.hasher.digest("hex");
+            resources.push({ name, path, bytes, hash });
+            logger.debug("Added file to WACZ", { path, bytes, hash }, "wacz");
+          }
+          currMarker = null;
+        } else if (isInFile) {
+          if (currMarker) {
+            yield chunk;
+            currMarker.hasher.update(chunk);
+            hasher.update(chunk);
+            size += chunk.length;
+          }
+        } else {
+          yield chunk;
+          hasher.update(chunk);
+          size += chunk.length;
+        }
+      }
+
+      wacz.hash = hasher.digest("hex");
+      wacz.size = size;
+    }
+
+    return Readable.from(iterWACZ(this));
+  }
+
+  getHash() {
+    return this.hash;
+  }
+
+  getSize() {
+    return this.size;
+  }
+
+  async generateToFile(filename: string) {
+    await pipeline(this.generate(), fs.createWriteStream(filename));
+  }
+
+  async *iterDirForZip(files: string[]): AsyncGenerator<InputWithoutMeta> {
+    const encoder = new TextEncoder();
+    const end = new EndMarker();
+    // correctly handles DST
+    const hoursOffset = (24 - new Date(0).getHours()) % 24;
+    const timezoneOffset = hoursOffset * 60 * 60 * 1000;
```

Comment:
As mentioned above, we might not need to do this at all actually.

Reply:
Looking at this again, the timezone if always UTC in the Docker container, unless changed otherwise. But ZIP file contents are generally treated as local timezone, unless an 'extended timestamp field' is used (which I don't think the client-zip library supports) or we pass the offset to the docker container somehow. The result is that when running on non-UTC host, the files in the WACZ do have the wrong timestamp. However, this is already the case with py-wacz, so maybe something we address separately if we want to address it. Otherwise, the WARCs in the WACZ do have a UTC timestamp always.

Reply:
Interesting. I didn't know that particular quirk of the ZIP format. Seems like maybe we should aim for parity with py-wacz/current behavior for now and create a separate issue to track fixing this on non-UTC hosts.

Reply:
yep, just removing the offset stuff altogether keeps that parity, did that for now.